kerchunk/grib2.py

```diff
@@ -582,3 +586,87 @@ def correct_hrrr_subhf_step(group: Dict) -> Dict:
     group["refs"]["step/0"] = enocded_val
 
     return group
+
+
+def parse_grib_idx(
+    fs: fsspec.AbstractFileSystem,
+    basename: str,
+    suffix: str = "idx",
+    tstamp: Optional[pd.Timestamp] = None,
+    validate: bool = False,
+) -> pd.DataFrame:
+    """
+    Standalone method used to extract metadata from a grib2 idx file(text) from NODD
+    :param fs: the file system to read from
+    :param basename: the base name is the full path to the grib file
+    :param suffix: the suffix is the ending for the idx file
+    :param tstamp: the timestamp to record for this index process
+    :return: the data frame containing the results
+    """
+
+    fname = f"{basename}.{suffix}"
+    fs.invalidate_cache(fname)
+    fs.invalidate_cache(basename)
+
+    baseinfo = fs.info(basename)
+
+    with fs.open(fname, "r") as f:
+        splits = []
+        for line in f.readlines():
+            try:
+                idx, offset, date, attrs = line.split(":", maxsplit=3)
+                splits.append([int(idx), int(offset), date, attrs])
+            except ValueError:
+                # Wrap the ValueError in a new one that includes the bad line
+                # If building the mapping, pick a different forecast run where the idx file is not broken
+                # If indexing a forecast using the mapping, fall back to reading the grib file
+                raise ValueError(f"Could not parse line: {line}")
+
+    result = pd.DataFrame(
+        data=splits,
+        columns=["idx", "offset", "date", "attrs"],
+    )
+
+    # Subtract the next offset to get the length using the filesize for the last value
+    result.loc[:, "length"] = (
+        result.offset.shift(periods=-1, fill_value=baseinfo["size"]) - result.offset
+    )
+
+    result.loc[:, "idx_uri"] = fname
+    result.loc[:, "grib_uri"] = basename
+    if tstamp is None:
+        tstamp = pd.Timestamp.now()
+    result.loc[:, "indexed_at"] = tstamp
+
+    if isinstance(fs, gcsfs.GCSFileSystem):
```

Comment:
Rather than having to import gcsfs (or s3fs), I recommend checking the fs.protocol or the existence of given fields in the info dict.

There is also an `fs.ukey()` method, which does what I think these blocks are intended for: test whether the remote file has changed or not. If you use this instead, you don't need the blocks at all. `.ukey` works for local files too.

Reply:
Why do I have to check whether the remote files are changed or not?

Reply:
But the code is just parsing the `.idx` files.

Reply:
Correct, so all the more reason not to require gcsfs

Reply:
What other filesystems should I cover?

Reply:
If you use `ukey`, you automatically cover them all. 

Question: why do we need these keys at all?

Reply:
Yes. Why do we need these keys? I used the `ukey` function. It is only generating a random string

Reply:
> It is only generating a random string. 

Are you sure it's _random_ ?

Reply:
No I'm not sure. But how can the `ukey` function can be used to check the Filesystem?

Reply:
> No I'm not sure.

```python
    def ukey(self, path):
        """Hash of file properties, to tell if it has changed"""
        return sha256(str(self.info(path)).encode()).hexdigest()
```